src/main/ptapi/io/piuio/lxio/lxio.c

```diff
@@ -0,0 +1,296 @@
+    lxio_light_state_buffer[4] |= (1 << 3);
+  }
+  if (ptapi_piubtn_out[1].right) {
+    lxio_light_state_buffer[4] |= (1 << 4);
+  }
+
+  // this should be player 1, but its player 2
+  if (ptapi_piubtn_out[0].back) {
+    lxio_light_state_buffer[5] |= (1 << 4);
+  }
+  if (ptapi_piubtn_out[0].start) {
+    lxio_light_state_buffer[5] |= (1 << 2);
+  }
+  if (ptapi_piubtn_out[0].left) {
+    lxio_light_state_buffer[5] |= (1 << 3);
+  }
+  if (ptapi_piubtn_out[0].right) {
+    lxio_light_state_buffer[5] |= (1 << 0);
+    lxio_light_state_buffer[5] |= (1 << 1);
+  }
+
+
+/**
+  for (uint8_t i = 0; i < 2; i++) {
+    //both UR and UL are mapped to "back"
+    if (ptapi_piubtn_out[i].back) {
+      lxio_light_state_buffer[i+4] |= (1 << 0);
+      lxio_light_state_buffer[i+4] |= (1 << 1);
+    }
+    if (ptapi_piubtn_out[i].start) {
+      lxio_light_state_buffer[i+4] |= (1 << 2);
+    }
+    if (ptapi_piubtn_out[i].left) {
+      lxio_light_state_buffer[i+4] |= (1 << 3);
+    }
+    if (ptapi_piubtn_out[i].right) {
+      lxio_light_state_buffer[i+4] |= (1 << 4);
+    }
+  }*/
+}
+
+static void convert_input_to_piuio()
+{
+  // LXIO sends all four sensor positions (the entire pad state 10*4) at once.
+  // Each byte is a sensor position, so we need to take a look at the bit flags.
+  // P1 is bytes 0->3 and P2 is 4->7 inclusive
+  // and LXIO follows the same order as PTAPI/PIUIO in sensor order
+  // that being lu, ru, cn, ld, rd (index: 0,1,2,3,4)
+  for (uint8_t i = 0; i < 2; i++) {
+    for (uint8_t j = 0; j < PTAPI_IO_PIUIO_SENSOR_GROUP_NUM; j++) {
+      ptapi_piuio_pad_in[i][j].lu = lxio_btn_state_buffer[i * 4 + j] & (1 << 0);
+      ptapi_piuio_pad_in[i][j].ru = lxio_btn_state_buffer[i * 4 + j] & (1 << 1);
+      ptapi_piuio_pad_in[i][j].cn = lxio_btn_state_buffer[i * 4 + j] & (1 << 2);
+      ptapi_piuio_pad_in[i][j].ld = lxio_btn_state_buffer[i * 4 + j] & (1 << 3);
+      ptapi_piuio_pad_in[i][j].rd = lxio_btn_state_buffer[i * 4 + j] & (1 << 4);
+    }
+  }
+
+  // cabinet buttons
+  ptapi_piuio_sys.test = lxio_btn_state_buffer[8] & (1 << 1);
+  ptapi_piuio_sys.service = lxio_btn_state_buffer[8] & (1 << 6);
+  ptapi_piuio_sys.clear = lxio_btn_state_buffer[8] & (1 << 7);
+  ptapi_piuio_sys.coin = lxio_btn_state_buffer[8] & (1 << 2);
+  ptapi_piuio_sys.coin2 = lxio_btn_state_buffer[9] & (1 << 2);
+
+}
+
+bool ptapi_io_piuio_recv(void)
+{
+  // NOTE: The lxio *MUST* be written to in order to be read.
+  // It cannot be read individually without a written payload. It will soft
+  // lock. In order to ensure the full cycle, everything is taken care of here.
+
+  // take the ptapi info and convert it to an lxio message.
+  convert_output_to_lxio();
+
+  if (!lxio_drv_device_write(lxio_light_state_buffer, LXIO_MSG_SIZE)) {
+    return false;
+  }
+
+  if (!lxio_drv_device_read(lxio_btn_state_buffer, LXIO_MSG_SIZE)) {
+    return false;
+  }
+
+  // TODO: lxio contains an unsigned 16bit packet counter in the last two LSB
```

Comment:
Want to leave that as an open TODO? Could also be left simply as a comment of something that is known but not implemented for given reasons.

Reply:
Left it as a note